asapdiscovery-ml/asapdiscovery/ml/inference/inference.py

```diff
@@ -0,0 +1,156 @@
+    """
+    Inference base class for PyTorch models in asapdiscovery.
+
+    Parameters
+    ----------
+    model_name : str
+        Name of model to use.
+    model_type : str
+        Type of model to use.
+    from_spec : bool, default=True
+        Whether to fetch weights from asapdiscovery.ml spec file or from local file.
+    build_model_kwargs : Optional[Dict], default=None
+        Keyword arguments to pass to build_model function.
+    device : str, default='cpu'
+        Device to use for inference.
+
+
+    Methods
+    -------
+    predict
+        Predict on a batch of data.
+
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        model_type: str,
+        model_spec: Path = None,
+        build_model_kwargs: Optional[dict] = None,
+        device: str = "cpu",
+    ):
+        logging.info(f"initializing {self.__class__.__name__} class")
+
+        self.device = device
+        logging.info(f"using device {self.device}")
+
+        self.model_name = model_name
+        self.model_type = model_type
+        self.model_spec = model_spec
+
+        self.model_components = None
+
+        logging.info(
+            f"using model {self.model_name} of type {self.model_type} from spec {self.model_spec}"
+        )
+
+        # load model weights or fetch them
+        if not self.model_spec:
+            logging.info(
+                " no model spec specified, using spec from asapdiscovery.ml models.yaml spec file"
+            )
+            self.model_spec = all_models
+        else:
+            logging.info("local yaml file specified, fetching weights from spec")
+            if not self.model_spec.split(".")[-1] in ["yaml", "yml"]:
+                raise ValueError(
+                    f"Model spec file {self.model_spec} is not a yaml file"
+                )
+
+        self.model_components = fetch_model_from_spec(self.model_spec, model_name)[
+            model_name
+        ]
+        if self.model_components.type != self.model_type:
+            raise ValueError(
+                f"Model type {self.model_components.type} does not match {self.model_type}"
+            )
+
+        logging.info(f"found weights {self.model_components.weights}")
+
+        # build model kwargs
+        if not build_model_kwargs:
+            build_model_kwargs = {}
+            build_model_kwargs["config"] = self.model_components.config
+
+        # otherwise just roll with what we have
+
+        # build model, this needs a bit of cleaning up in the function itself.
+        self.model = self.build_model(self.model_components.type, **build_model_kwargs)
+        logging.info(f"built model {self.model}")
+
+        # load weights
+        self.model = load_weights(self.model, self.model_components.weights)
+        logging.info(f"loaded weights {self.model_components.weights}")
+
+        self.model.eval()
+        logging.info("set model to eval mode")
+
+    def build_model(self, model_type: str, **kwargs):
```

Comment:
should we also include `*args` here or is there not really any reason for it?

Reply:
Other than `model_type` the rest of the args to `util.build_model` are `kwargs` so I don't see a reason. If we want we can always change it easily down the track.